### ISSUE TITLE
Agones controller shut down 

### ISSUE THREAD

**Issue by KamiMay** [googleforgames/agones#1170]
**What happened**: We have experienced cluster wide down time after Game Server list got out of sync with Pod list. When investigating Agones controller logs I can see that it has shut down, but has not started back up for 12 minutes after shutting down. 

Logs:
```
2019-11-02 23:28:37.132 GMT Shut down agones controllers
2019-11-02 23:40:19.201 GMT logging to file
2019-11-02 23:40:19.202 GMT starting gameServer operator...
2019-11-02 23:40:19.208 GMT Starting
```

After this game server list was extremely out of sync with actual pod list. 
Running `kubectl get gs` returned:
```
NAME                                STATE       ADDRESS      PORT   NODE                          AGE
deathmatch-server-odd-4fvpv-fjl8q   Ready       34.87.124.   7014   gke-node-pool-157ec1db-kbq3   8h
deathmatch-server-odd-4fvpv-g2dn2   Ready       34.87.79.    7534   gke-node-pool-55d97c47-dl00   8h
deathmatch-server-odd-4fvpv-h28gr   Allocated   35.247.128   7112   gke-node-pool-157ec1db-mgh3   8h
deathmatch-server-odd-4fvpv-hcwdv   Allocated   34.87.124.   7265   gke-node-pool-157ec1db-kbq3   8h
deathmatch-server-odd-4fvpv-hwgps   Allocated   35.247.128   7782   gke-node-pool-157ec1db-mgh3   8h
deathmatch-server-odd-4fvpv-mrcgx   Allocated   34.87.124.   7111   gke-node-pool-157ec1db-kbq3   8h
deathmatch-server-odd-4fvpv-nq58m   Ready       35.247.128   7861   gke-node-pool-157ec1db-mgh3   8h
deathmatch-server-odd-4fvpv-s765b   Ready       34.87.79.    7095   gke-node-pool-55d97c47-dl00   8h
deathmatch-server-odd-4fvpv-wrzlv   Ready       34.87.124.   7966   gke-node-pool-157ec1db-kbq3   8h
endless-server-odd-h79b4-28j55      Allocated   35.247.128   7601   gke-node-pool-157ec1db-mgh3   9h
endless-server-odd-h79b4-5q8ws      Allocated   35.247.17    7990   gke-node-pool-55d97c47-f2nv   20h
endless-server-odd-h79b4-qshqv      Ready       34.87.124.   7612   gke-node-pool-157ec1db-kbq3   8h
lobby-server-odd-znvjb-7vmbq        Allocated   35.187.246   7148   gke-node-pool-fb3f8ee9-rvpw   7s
lobby-server-odd-znvjb-dzpfl        Allocated   35.247.128   7992   gke-node-pool-157ec1db-mgh3   8h
lobby-server-odd-znvjb-fbfcr        Allocated   35.247.128   7209   gke-node-pool-157ec1db-mgh3   8h
lobby-server-odd-znvjb-lmlqt        Allocated   35.187.246   7167   gke-node-pool-fb3f8ee9-rvpw   15s
lobby-server-odd-znvjb-nnbnc        Allocated   35.247.128   7294   gke-node-pool-157ec1db-mgh3   8h
lobby-server-odd-znvjb-nrmj7        Allocated   35.247.128   7555   gke-node-pool-157ec1db-mgh3   8h
lobby-server-odd-znvjb-s2nkj        Ready       34.87.15.    7645   gke-node-pool-55d97c47-67s0   19s
lobby-server-odd-znvjb-skk6b        Allocated   35.247.128   7848   gke-node-pool-157ec1db-mgh3   13h
lobby-server-odd-znvjb-stqm2        Allocated   35.247.128   7193   gke-node-pool-157ec1db-mgh3   13h
lobby-server-odd-znvjb-t8m2z        Allocated   35.247.128   7220   gke-node-pool-157ec1db-mgh3   8h
lobby-server-odd-znvjb-tf5wh        Allocated   35.247.128   7037   gke-node-pool-157ec1db-mgh3   10h
lobby-server-odd-znvjb-tnfvb        Allocated   35.247.128   7824   gke-node-pool-157ec1db-mgh3   10h
lobby-server-odd-znvjb-tzq8f        Allocated   35.187.246   7639   gke-node-pool-fb3f8ee9-rvpw   31s
lobby-server-odd-znvjb-v4qgl        Allocated   35.247.128   7292   gke-node-pool-157ec1db-mgh3   8h
lobby-server-odd-znvjb-v8gzf        Allocated   35.247.128   7679   gke-node-pool-157ec1db-mgh3   13h
lobby-server-odd-znvjb-vksdf        Allocated   35.247.128   7101   gke-node-pool-157ec1db-mgh3   13h
lobby-server-odd-znvjb-vshkb        Ready       34.87.124.   7660   gke-node-pool-157ec1db-kbq3   43s
lobby-server-odd-znvjb-vtg4z        Allocated   35.247.128   7907   gke-node-pool-157ec1db-mgh3   13h
lobby-server-odd-znvjb-vvnkz        Allocated   35.247.128   7832   gke-node-pool-157ec1db-mgh3   12h
lobby-server-odd-znvjb-w2kc4        Ready       35.240.183   7388   gke-node-pool-fb3f8ee9-lw1p   56s
lobby-server-odd-znvjb-w6sdr        Allocated   35.247.128   7532   gke-node-pool-157ec1db-mgh3   9h
lobby-server-odd-znvjb-wf944        Allocated   35.247.128   7681   gke-node-pool-157ec1db-mgh3   16h
lobby-server-odd-znvjb-x5hpd        Ready       35.240.183   7149   gke-node-pool-fb3f8ee9-lw1p   10m
lobby-server-odd-znvjb-xdlh6        Allocated   35.247.128   7763   gke-node-pool-157ec1db-mgh3   14h
lobby-server-odd-znvjb-xjnzl        Ready       35.240.183   7372   gke-node-pool-fb3f8ee9-lw1p   19m
lobby-server-odd-znvjb-xp7h4        Ready       35.240.183   7847   gke-node-pool-fb3f8ee9-lw1p   17m
lobby-server-odd-znvjb-xp8fz        Allocated   35.247.128   7250   gke-node-pool-157ec1db-mgh3   14h
lobby-server-odd-znvjb-xrmx9        Ready       35.240.183   7085   gke-node-pool-fb3f8ee9-lw1p   39m
lobby-server-odd-znvjb-xs57n        Ready       35.240.183   7999   gke-node-pool-fb3f8ee9-lw1p   22m
lobby-server-odd-znvjb-z6ptt        Ready       35.240.183   7949   gke-node-pool-fb3f8ee9-lw1p   6m
lobby-server-odd-znvjb-zbwnr        Ready       35.240.183   7171   gke-node-pool-fb3f8ee9-lw1p   16m
lobby-server-odd-znvjb-zc8wl        Ready       35.240.183   7269   gke-node-pool-fb3f8ee9-lw1p   20m
lobby-server-odd-znvjb-zcx2r        Ready       35.240.183   7046   gke-node-pool-fb3f8ee9-lw1p   1m
lobby-server-odd-znvjb-zhdw8        Ready       35.240.183   7315   gke-node-pool-fb3f8ee9-lw1p   14m
lobby-server-odd-znvjb-zjrtd        Ready       35.240.183   7080   gke-node-pool-fb3f8ee9-lw1p   13m
lobby-server-odd-znvjb-zkvfb        Ready       35.247.128   7662   gke-node-pool-157ec1db-mgh3   9h
lobby-server-odd-znvjb-zpffl        Ready       35.240.183   7396   gke-node-pool-fb3f8ee9-lw1p   11s
lobby-server-odd-znvjb-zs5c4        Ready       35.240.183   7956   gke-node-pool-fb3f8ee9-lw1p   1m
lobby-server-odd-znvjb-ztn96        Ready       35.240.183   7397   gke-node-pool-fb3f8ee9-lw1p   32m
lobby-server-odd-znvjb-zv697        Ready       35.247.128   7377   gke-node-pool-157ec1db-mgh3   14h
lobby-server-odd-znvjb-zvjrl        Ready       35.240.183   7878   gke-node-pool-fb3f8ee9-lw1p   23m
lobby-server-odd-znvjb-zzftl        Ready       35.247.128   7835   gke-node-pool-157ec1db-mgh3   16h
messenger-server-odd-8l7ng-fv64c    Ready       34.87.79.    7036   gke-node-pool-55d97c47-dl00   14h
messenger-server-odd-8l7ng-fv64c    Ready       34.87.79.    7036   gke-node-pool-55d97c47-dl00   14h
royale-server-odd-d5wnz-8dfsq       Ready       35.240.183   7393   gke-node-pool-fb3f8ee9-lw1p   5m
royale-server-odd-d5wnz-96q8h       Shutdown    35.187.246   7004   gke-node-pool-fb3f8ee9-rvpw   11m
royale-server-odd-d5wnz-b4877       Ready       35.240.183   7081   gke-node-pool-fb3f8ee9-lw1p   14m
royale-server-odd-d5wnz-cw76d       Ready       35.240.183   7923   gke-node-pool-fb3f8ee9-lw1p   3m
royale-server-odd-d5wnz-d45mk       Ready       35.240.183   7732   gke-node-pool-fb3f8ee9-lw1p   21m
royale-server-odd-d5wnz-dx9vw       Allocated   34.87.124.   7266   gke-node-pool-157ec1db-kbq3   8h
royale-server-odd-d5wnz-qxgmz       Allocated   35.247.128   7202   gke-node-pool-157ec1db-mgh3   8h
royale-server-odd-d5wnz-wtbkz       Ready       34.87.124.   7495   gke-node-pool-157ec1db-kbq3   7s
```

And running `kubectl get pods` returned:
```
NAME                                READY   STATUS        RESTARTS   AGE
deathmatch-server-odd-4fvpv-fjl8q   2/2     Running       0          8h
deathmatch-server-odd-4fvpv-g2dn2   2/2     Running       0          8h
deathmatch-server-odd-4fvpv-nq58m   2/2     Running       0          8h
deathmatch-server-odd-4fvpv-s765b   2/2     Running       5          8h
endless-server-odd-h79b4-5q8ws      2/2     Running       0          20h
endless-server-odd-h79b4-qshqv      2/2     Running       0          8h
exchange-odd-77475599bf-rt77n       1/1     Running       0          14h
lobby-server-odd-znvjb-6wk64        2/2     Running       0          5s
lobby-server-odd-znvjb-7vmbq        2/2     Running       0          30s
lobby-server-odd-znvjb-lmlqt        0/2     Terminating   0          38s
lobby-server-odd-znvjb-tzq8f        2/2     Running       0          54s
lobby-server-odd-znvjb-vshkb        2/2     Running       0          66s
lobby-server-odd-znvjb-w2kc4        2/2     Running       0          79s
lobby-server-odd-znvjb-x5hpd        2/2     Running       0          10m
lobby-server-odd-znvjb-xjnzl        2/2     Running       0          20m
lobby-server-odd-znvjb-xp7h4        2/2     Running       0          17m
lobby-server-odd-znvjb-xrmx9        2/2     Running       0          39m
lobby-server-odd-znvjb-xs57n        2/2     Running       0          22m
lobby-server-odd-znvjb-z6ptt        2/2     Running       0          6m52s
lobby-server-odd-znvjb-zbwnr        2/2     Running       0          16m
lobby-server-odd-znvjb-zc8wl        2/2     Running       0          20m
lobby-server-odd-znvjb-zcx2r        2/2     Running       0          90s
lobby-server-odd-znvjb-zhdw8        2/2     Running       0          15m
lobby-server-odd-znvjb-zjrtd        2/2     Running       0          13m
lobby-server-odd-znvjb-zpffl        2/2     Running       0          34s
lobby-server-odd-znvjb-zs5c4        2/2     Running       0          2m15s
lobby-server-odd-znvjb-ztn96        2/2     Running       0          33m
lobby-server-odd-znvjb-zvjrl        2/2     Running       0          23m
matchmaker-odd-54d4cc75f-8sxb5      1/1     Running       0          8h
messenger-server-odd-8l7ng-fv64c    2/2     Running       0          14h
progression-odd-d46b8674-jkn65      1/1     Running       0          8h
royale-server-odd-d5wnz-8dfsq       2/2     Running       0          2m50s
royale-server-odd-d5wnz-96q8h       2/2     Running       0          9m16s
royale-server-odd-d5wnz-b4877       2/2     Running       0          12m
royale-server-odd-d5wnz-cw76d       2/2     Running       0          75s
royale-server-odd-d5wnz-d45mk       2/2     Running       0          18m
royale-server-odd-d5wnz-t6hf5       2/2     Running       0          25m
```

Unsure, but maybe this might be related: https://github.com/googleforgames/agones/issues/398

**What you expected to happen**: I expected for this to never happen.

**How to reproduce it (as minimally and precisely as possible)**: Unsure as to why Agones controller shut down and why it took this long to restart. so Unsure on how to reproduce. 

**Environment**:
- Agones version:  1.0
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.11-dispatcher", GitCommit:"2e298c7e992f83f47af60cf4830b11c7370f6668", GitTreeState:"clean", BuildDate:"2019-09-19T22:20:12Z", GoVersion:"go1.11.13", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"12+", GitVersion:"v1.12.10-gke.15", GitCommit:"7b5157a7c600aa8ee6e2b36b56c478e030d5bfe2", GitTreeState:"clean", BuildDate:"2019-10-07T20:39:12Z", GoVersion:"go1.11.13b4", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: GCP n1-standard-1
- Install method (yaml/helm): helm

**Comment by markmandel** [comment]
Is there anything in the logs about pod evictions, or node updates, or resource issues anything like that?

Where the health checks failing on the controller?

How did you restart it, once it had fallen over?

**Comment by KamiMay** [comment]
It is quite possible master node has been updated, but I am not sure how to check when it happened, node upgrades are disabled, so nodes should not be affected by kubernetes upgrades. Looking at logs it looks healthy, the only odd thing is `Enqueuing` spam. This is screenshot of logs just before it shut down.

![image](https://user-images.githubusercontent.com/2967343/68142765-f6be8800-ff27-11e9-92f7-3b7ae6d2c127.png)

No health checks seem to be failing, the only errors I can see are `error updating GameServer  to Scheduled state: Operation cannot be fulfilled on gameservers.agones.dev "SERVER NAME": the object has been modified; please apply your changes to the latest version and try again`

It restarted self 12 mins after as we can see in the logs, I was made aware of this incident only 8 hours after it happened. I fixed it by deleting game servers that didn't have corresponding pods. That synced game servers and pods and players could continue playing.

**Comment by markmandel** [comment]
Weird. It isn't like it crashed, it looks like it shutdown gracefully. It almost would make more sense if it did some kind of hard crash.

Interesting that it self restarted - makes me think there was some kind of master upgrade - but that shouldn't have cause Agones to shutdown. I mean the controller is a simple K8s deployment -- no extra magic there. So if it crashes, it should come back up pretty quick.

:thinking: 

Anything in the K8s API logs? (I'm grasping for straws at this point)

**Comment by markmandel** [comment]
I think a good first step to have e2e tests, where we start up some game servers, shut down the controller, delete some things (pods, gameservers), and then startup the controller again -- and make sure we end up back in a healthy state.

**Comment by aLekSer** [comment]
A way to reproduce an issue with deleting controller could be next:
```
cd ./build
make install
kubectl apply -f ../examples/simple-udp/fleet.yaml
kubectl delete deployment  agones-controller  --namespace agones-system
kubectl delete pods
make install
```
With resulting resources:
```
kubectl get gs | wc -l 
2
kubectl get pods | wc -l 
0
```

**Comment by KamiMay** [comment]
We had this symptom happen again due to a different issue. 
We observed a critical error `2020-01-11 09:09:07.885 GMT Could not run sidecar`
Straight after there was a string of `Failed to watch` errors eg:

```
agones.dev/agones/pkg/client/informers/externalversions/factory.go:119: Failed to watch *v1.GameServer: Get https://10.0.0.1:443/apis/agones.dev/v1/namespaces/default/gameservers?fieldSelector=metadata.name%3Dlobby-server-tk4dn-rmsvl&resourceVersion=112789068&timeoutSeconds=317&watch=true: dial tcp 10.0.0.1:443: i/o timeout
```

Game servers managed to get out of sync and we had a bunch gs that were no longer server by any pods. This caused about 36h downtime in one region due to happening during weekend and went unnoticed because it was appearing as healthy on the server. 

Currently I will work on a workaround, I will make a self healing service that will periodically check that all gs are being served by pods and delete any that managed to get out of sync. But preferably agones would take care of it.

**Comment by markmandel** [comment]
(We talked about this a bit on chat) -- while we should check for gameservers that don't have pods (probably ones that are after ready), and move them to Unhealthy on startup? (or maybe continuously?? They should be caught by the health manager). I'm curious -- why are your pods being shut down while the controller isn't working?

Are you running a process.exit() in your gameserver?

I'm wondering if the answer is rather than have the process exit itself, but call SDK.Shutdown(), which will succeed when master comes back up / the issue resolves itself?

And then the gameserver binary will be stopped by the controller, rather than exiting itself.

Does that make sense?

**Comment by markmandel** [comment]
After doing the PVM Fix (#1245) a similar solution should work for this as well.

I _think_ using the SDK for Shutdown() always should solve this most of the time -- _but_ sometimes weird things happens, and we need to eventually heal.
